Add Slab2 model to Rockhound

aguspesce opened the pull request:
Add function to fetch and load data from Slab2: a geometry model of
subduction zones, provided by the [USGS](https://www.sciencebase.gov/catalog/item/5aa1b00ee4b0b1c392e86467). The model contains gridded data
for 27 subduction zones describing the depth, thickness, dip and strike
angles of the subducted slab. The data for each subduction zone is
fetched and loaded into a `xarray.Dataset`. The depth and thickness are
converted to meters and the coordinates renamed to "longitude" and
"latitude". Add metadata to the Dataset like the datum and the doi of
the model. Add netcdf4 as a requirement for Rockhound.

Fixes #61

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [x] Add new public functions/methods/classes to `doc/api/index.rst`.
- [x] Write detailed docstrings for all functions/methods.
- [x] If adding new functionality, add an example to the docstring, gallery, and/or tutorials.

aguspesce committed:
Add only 2 subduction zone of the Slab2 model to registry file.

As a first step, I add the names, hashes and download directions
of the Alaska and Calabria subduction zones taken from the Slab2 model.

aguspesce committed:
Add netcdf dependence in the environment file.

aguspesce committed:
Add fetch_slab2 function to the init file

aguspesce committed:
Create a function to fetch the Slab2 model.

For now, this function only fetch only 2 subduction zones.

aguspesce commented:
@leouieda  and @santisoler,
I add in the registry file one by one the geometry files for each zone.
I do that because the USGS allows to download a zip file for a given zone but the hash in its name change every time that I download this file.

Do you have other idea to improve this problem?

aguspesce committed:
Improve the documentation of slab2 function and add its reference.

aguspesce committed:
Eliminate the possibility to choose which data to download

aguspesce committed:
The dataset units are change from km to meters

aguspesce committed:
Fix the Slab2 reference

aguspesce committed:
Create the example to use the Slab2 model

aguspesce committed:
Add different attributes, units and name to the Slab2 dataset

santisoler commented:
Happy to see this PR, @aguspesce!

I think it would be ideal to have a single zip file for each region on Slab2 model, but if it's not possible to have a consistent zip file in time, I think it's better to just add a ton of files to the registry.
I don't think this is a big issue. The registry file is not intended to be readable, so I don't see any problem with having a large number of files in it.

Ask for review whenever you are ready, or for help if you need it.

aguspesce committed:
Add all subduction zone of the Slab2 model to registry file.

aguspesce committed:
Add all Slab2 zones and fix bad registry urls

Fix bad ulrs on the registry.txt file and populate the ZONES dictionary
with the rest of the zones available on Slab2.

aguspesce committed:
Fix error message for invalid zone

aguspesce committed:
Improve the example of slab2.py using cartopy

aguspesce committed:
Add the available zones in the documentation of slab2 function.

aguspesce commented:
@santisoler  and @leouieda, 
The function to fetch Slab2 model  is almost finished but I need your help to do the testing functions.

santisoler commented on examples/slab2.py:
```diff
@@ -0,0 +1,32 @@
+import cartopy.crs as ccrs
+
+# Load a geometry of the Alaska subduction zone
+grid = rh.fetch_slab2(zone="south_america")
```
The zone in the comment does not corresponds with the actual fetched zone.

aguspesce replied:
Done

santisoler reviewed (changes requested):
This is looking great @aguspesce! :1st_place_medal: 
Just found a little mistake on the example and left a few comments on the code, nothing to worry about.
I'll try to create some test functions to inspire you.

santisoler commented on rockhound/slab2.py:
```diff
@@ -0,0 +1,141 @@
+    ds.attrs.update(
+        {
+            "title": "Slab2 model - Zone: {}".format(ZONES[zone]["name"]),
+            "zone": ZONES[zone]["name"],
```
I think would be nice to have save the passed `zone` argument into the `zone` attribute of the `xr.Dataset`, but also is a good idea to save the full name of the zone.
So, maybe we could change this like:
```suggestion
            "zone": zone,
            "zone_full_name": ZONES[zone]["name"],
```

santisoler commented on rockhound/slab2.py:
```diff
@@ -0,0 +1,141 @@
+            "title": "Slab2 model - Zone: {}".format(ZONES[zone]["name"]),
+            "zone": ZONES[zone]["name"],
+            "datum": "WGS84",
+            "doi": "10.5066/F7PV6JNV.",
```
Remove the final dot:
```suggestion
            "doi": "10.5066/F7PV6JNV",
```

santisoler commented on rockhound/slab2.py:
```diff
@@ -0,0 +1,141 @@
+        array.attrs["units"] = DATASETS[dataset]["units"]
+    # Merge arrays into a single xr.Dataset
+    ds = xr.merge(arrays)
+    # Change units of thickness, depth and depth_uncertainty to meters
```
We also need to modify the units of the `actual_range` attribute of the `thickness`, `depth` and `depth_uncertainty`. They are currently in the original units (km) what does not agree with the actual units of the data.

aguspesce replied:
I resolved this!

santisoler committed:
Add test functions for Slab2 fetch function

santisoler commented:
I've just pushed a commit that adds test functions for Slab2. They will probably fail due to some differences between the expected values of the DOI and the `actual_range` attributes of `depth`, `thickness` and `depth_uncertainty` (see my previous comments). Don't worry, after they are fixed, the tests should pass.

In case you like my suggestion of adding the actual `zone` argument to the `zone` attribute, one test function must be updated.

santisoler committed:
Rename ds to dataset to improve styling

aguspesce committed:
Update the attribute of the zone name in slab2.py

Co-Authored-By: Santiago Soler <santiago.r.soler@gmail.com>

aguspesce committed:
Fix the doi  in slab2.py

Co-Authored-By: Santiago Soler <santiago.r.soler@gmail.com>

aguspesce committed:
Change unit of actual_range attribute

aguspesce committed:
Plot all subduction zones in a map

aguspesce committed:
Fix the format in the slab2 example

aguspesce commented:
Hello @santisoler, 
I make some changes... Place, see them

santisoler commented on examples/slab2.py:
```diff
@@ -13,19 +13,48 @@
+    "sulawesi",
+    "sumatra_java",
+    "vanuatu",
+}
```
There's no need to rewrite every zone available on Slab2.
You can import the `ZONES` dictionary directly and then iterate over its keys:
```python
from rockhound.slab2 import ZONES

...

for zone in ZONES:
    grid = rh.fetch_slab2(zone)
    ...
```

aguspesce replied:
Done!

santisoler reviewed (changes requested):
Just a suggestion to avoid repeating code.

aguspesce committed:
Add the check of the zone and the long name on test_slab2

aguspesce committed:
Run make lint and delete unused argument and os package

aguspesce committed:
Improve the slab2 example.

aguspesce committed:
Fix wrong citation of Slab2

aguspesce committed:
Compare min max values with actual_range using assert_allclose

aguspesce committed:
Add fetch_slab2 to doc/api/index.rst.

santisoler commented on rockhound/slab2.py:
```diff
@@ -0,0 +1,147 @@
+            - `sumatra_java`: Sumatra-Java
+            - `vanuatu`: Vanuatu
+    load : bool
+        Wether to load the data into an :class:`xarray.Dataset` or just return the
```
```suggestion
        Whether to load the data into an :class:`xarray.Dataset` or just return the
```

santisoler committed:
Add netcdf4 to requirements.txt

santisoler commented on rockhound/slab2.py:
```diff
@@ -0,0 +1,147 @@
+    grid.depth.attrs["actual_range"] = grid.depth.attrs["actual_range"] * 1000
+    grid.thickness.attrs["actual_range"] = grid.thickness.attrs["actual_range"] * 1000
+    grid.depth_uncertainty.attrs["actual_range"] = (
+        grid.depth_uncertainty.attrs["actual_range"] * 1000
```
This should be on the same place where the other units are changed, just to improve readability.

santisoler committed:
Simplify units conversion of fields and actual_range

santisoler committed:
Fix indentation and rst style of zones list in docstring

santisoler committed:
Merge branch 'master' into slab2-model

aguspesce committed:
Fix doc string in the documentation

Co-Authored-By: Santiago Soler <santiago.r.soler@gmail.com>

santisoler commented on doc/references.rst:
```diff
@@ -6,3 +6,4 @@ References
 .. [IRIS2011] IRIS DMC (2011), Data Services Products: EMC, A repository of Earth models, doi:10.17611/DP/EMC.1.
 .. [BEDMAP2] Fretwell, P. et al. (2013). Bedmap2: improved ice bed, surface and thickness datasets for Antarctica. The Cryosphere. doi:10.5194/tc-7-375-2013
 .. [Muller2008] Müller, R. D., Sdrolias, M., Gaina, C., & Roest, W. R. (2008). Age, spreading rates, and spreading asymmetry of the world’s ocean crust. Geochemistry, Geophysics, Geosystems, 9(4). doi:10.1029/2007GC001743
+.. [SLAB2] Hayes, G., 2018, Slab2 - A Comprehensive Subduction Zone Geometry Model: U.S. Geological Survey data release, https://doi.org/10.5066/F7PV6JNV.
```
```suggestion
.. [SLAB2] Hayes, G. (2018). Slab2 - A Comprehensive Subduction Zone Geometry Model: U.S. Geological Survey data release, https://doi.org/10.5066/F7PV6JNV.
```

santisoler commented on examples/slab2.py:
```diff
@@ -0,0 +1,39 @@
+        transform=ccrs.PlateCarree(),
+        add_colorbar=False,
+        vmin=-701160.33935547,
+        vmax=-1399.69277382,
```
Instead of hardcoding the min and max values I would compute them inside the example.

santisoler commented on examples/slab2.py:
```diff
@@ -0,0 +1,39 @@
+        vmax=-1399.69277382,
+    )
+
+ax.set_title("Slab2 - A comprehensive subduction zone geometry model")
```
```suggestion
ax.set_title("Slab2: Geometry model for subduction zones")
```

santisoler reviewed (changes requested):
This is looking great @aguspesce and I think we are very close to merge it :medal_sports:.

I would only improve some little things on the example.
Would you add some comments explaining what are you doing? The idea is that any new user could see a simple example on how this dataset could be loaded and plotted. So some comments are very welcome.

santisoler committed:
Compute min and max values inside the example

Instead of hardcoding the min and max values of the subducting plates
depths, it's better to compute them on the example.

aguspesce committed:
Improve the title of subducting plate plot

Co-Authored-By: Santiago Soler <santiago.r.soler@gmail.com>

aguspesce committed:
Inprove the slab2 reference

Co-Authored-By: Santiago Soler <santiago.r.soler@gmail.com>

aguspesce committed:
Refactor the examples

santisoler committed:
Shrink colorbar on slab2 example

santisoler commented:
This is looking awesome @aguspesce! Congratulations for the hard work!
I've just shrinked the colorbar on the example just to improve how the plot looks.
After the CIs pass I think this is ready to be merged.
